### PR TITLE
Add GrpcExecutors metrcis

### PR DESCRIPTION
**Pull request by xichen01**
### What changes are proposed in this pull request?
Added some GrpcExecutors ThreadPool metrics

### Why are the changes needed?
Add a way to monitor worker load info. Include worker read/write/cache thread pool.

### Does this PR introduce any user facing changes?

no

### CODE REVIEW

**Review by maobaolong — COMMENTED** [review]
@xichen01 Thanks for contributing these useful metrics, it really make it easy to know the worker status through these metrics.

Please take a double check is there any expensive operation of each metric？

**Comment by xichen01** [comment]
The time complexity of metrics operation is O(1), so it is a cheap operation

**Review by maobaolong — APPROVED** [review]
Thanks for your contribution, these metrics look worth to have

File: core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java [hunk 1]
```diff
@@ -70,12 +71,31 @@
           new ImpersonateThreadPoolExecutor(BLOCK_WRITE_THREAD_POOL_EXECUTOR);
 
   static {
+    MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
+            MetricKey.WORKER_CACHE_MANAGER_THREAD_ACTIVE_COUNT.getName()),
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR::getActiveCount);
+    MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
+            MetricKey.WORKER_CACHE_MANAGER_THREAD_CURRENT_COUNT.getName()),
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR::getPoolSize);
```
Comment (jiacheliu3) [inline core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java:79]: What's the difference between these two?

File: core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java [hunk 2]
```diff
@@ -70,12 +71,31 @@
...
+    MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
+            MetricKey.WORKER_CACHE_MANAGER_THREAD_QUEUE_WAITING_TASK_COUNT.getName()),
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR.getQueue()::size);
+    MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
```
Comment (jiacheliu3) [inline core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java:83]: I think this is a static value that does not change. So there's no need to monitor it.

File: core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java [hunk 3]
```diff
@@ -70,12 +71,31 @@
...
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR::getCompletedTaskCount);
+
     MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
         MetricKey.WORKER_BLOCK_READER_THREAD_ACTIVE_COUNT.getName()),
         BLOCK_READER_THREAD_POOL_EXECUTOR::getActiveCount);
     MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
         MetricKey.WORKER_BLOCK_READER_THREAD_CURRENT_COUNT.getName()),
         BLOCK_READER_THREAD_POOL_EXECUTOR::getPoolSize);
+    MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
+            MetricKey.WORKER_BLOCK_READER_THREAD_QUEUE_WAITING_TASK_COUNT.getName()),
```
Comment (jiacheliu3) [inline core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java:97]: same indentation as other lines

File: core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java [hunk 4]
```diff
@@ -70,12 +71,31 @@
...
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR::getPoolSize);
+    MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
+            MetricKey.WORKER_CACHE_MANAGER_THREAD_QUEUE_WAITING_TASK_COUNT.getName()),
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR.getQueue()::size);
+    MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
+            MetricKey.WORKER_CACHE_MANAGER_THREAD_MAX_COUNT.getName()),
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR::getMaximumPoolSize);
+    MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
+            MetricKey.WORKER_CACHE_MANAGER_COMPLETED_TASK_COUNT.getName()),
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR::getCompletedTaskCount);
```
Comment (jiacheliu3) [inline core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java:88]: nit: Could you try to make the indentation and line wrapping identical for all these lines?

File: core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java [hunk 5]
```diff
@@ -38,13 +38,14 @@
...
 
-  public static final ExecutorService CACHE_MANAGER_EXECUTOR =
-      new ImpersonateThreadPoolExecutor(new ThreadPoolExecutor(THREADS_MIN,
+  public static final ThreadPoolExecutor CACHE_MANAGER_THREAD_POOL_EXECUTOR =
+      new ThreadPoolExecutor(THREADS_MIN,
           ServerConfiguration.getInt(PropertyKey.WORKER_NETWORK_ASYNC_CACHE_MANAGER_THREADS_MAX),
-          THREAD_STOP_MS, TimeUnit.MILLISECONDS,
-          new UniqueBlockingQueue<>(ServerConfiguration.getInt(
-              PropertyKey.WORKER_NETWORK_ASYNC_CACHE_MANAGER_QUEUE_MAX)),
-          ThreadFactoryUtils.build("CacheManagerExecutor-%d", true)));
```
Comment (jiacheliu3) [inline core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java:47]: nit: restore indentation if there's no change in these lines

File: core/common/src/main/java/alluxio/metrics/MetricKey.java [hunk 6]
```diff
@@ -1353,6 +1387,13 @@ public MetricKey build() {
           .setMetricType(MetricType.GAUGE)
           .setIsClusterAggregated(false)
           .build();
+  public static final MetricKey WORKER_BLOCK_READER_THREAD_QUEUE_WAITING_TASK_COUNT =
+      new Builder("Worker.BlockReaderThreadQueueWaitingTaskCount")
+          .setDescription("The current number of tasks waiting in the work queue "
+              + "in the reader thread thread pool")
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/metrics/MetricKey.java:1393]:
```suggestion
          .setDescription("The current number of tasks waiting in the work queue "
              + "in the reader thread thread pool, bounded by " + PropertyKey.WORKER_NETWORK_ASYNC_CACHE_MANAGER_QUEUE_MAX)
```

File: core/common/src/main/java/alluxio/metrics/MetricKey.java [hunk 7]
```diff
@@ -1380,6 +1421,13 @@ public MetricKey build() {
           .setMetricType(MetricType.GAUGE)
           .setIsClusterAggregated(false)
           .build();
+  public static final MetricKey WORKER_BLOCK_WRITER_THREAD_QUEUE_WAITING_TASK_COUNT =
+      new Builder("Worker.BlockWriterThreadQueueWaitingTaskCount")
+          .setDescription("The current number of tasks waiting in the work queue "
+              + "in the block write thread pool")
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/metrics/MetricKey.java:1427]:
```suggestion
              + "in the block writer thread pool")
```
In this is bounded by a property key, pls include in the description as above

File: core/common/src/main/java/alluxio/metrics/MetricKey.java [hunk 8]
```diff
@@ -1340,6 +1340,40 @@ public MetricKey build() {
...
+          .build();
+  public static final MetricKey WORKER_CACHE_MANAGER_THREAD_CURRENT_COUNT =
+      new Builder("Worker.CacheMangerThreadCurrentCount")
+          .setDescription("The current number of cache threads in the cache manger thread pool")
+          .setMetricType(MetricType.GAUGE)
+          .setIsClusterAggregated(false)
+          .build();
+  public static final MetricKey WORKER_CACHE_MANAGER_THREAD_QUEUE_WAITING_TASK_COUNT =
+      new Builder("Worker.CacheMangerThreadQueueWaitingTaskCount")
+          .setDescription("The current number of tasks waiting in the work queue "
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/metrics/MetricKey.java:1358]: Is this bounded by a property key? If so pls include that in the description.

File: core/common/src/main/java/alluxio/metrics/MetricKey.java [hunk 9]
```diff
@@ -1340,6 +1340,40 @@ public MetricKey build() {
...
+              + "in the cache manager thread pool")
+          .setMetricType(MetricType.GAUGE)
+          .setIsClusterAggregated(false)
+          .build();
+  public static final MetricKey WORKER_CACHE_MANAGER_THREAD_MAX_COUNT =
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/metrics/MetricKey.java:1363]: If this is actually bounded by a property key, i don't think we need to look at this?

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Mostly LGTM. My comments are mostly about formating. Thanks for the improvements!

My high level comment is, I can see you are trying to add metrics on all attributes of thread pools. Did you monitor the values of these metrics in your test environment? I think maybe some fields can be merged so that every metric means something, instead of we have all sorts of metrics but don't know which to look at.

In terms of performance, at this moment I'm not worried about monitoring many metrics at the worker because it's rarely an issue for the worker to serve metrics. But in general, I believe a good practice is to cache gauges as much as possible. If you cache some(if not all) of your gauges by a few seconds, that will be great.

File: core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java (hunk 2 above)
Comment (jiacheliu3) [reply core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java:83]: see above

File: core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java [hunk 10]
```diff
@@ -70,6 +71,22 @@ public final class GrpcExecutors {
...
 
   static {
+    MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
+        MetricKey.WORKER_CACHE_MANAGER_THREAD_ACTIVE_COUNT.getName()),
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR::getActiveCount);
+    MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
+        MetricKey.WORKER_CACHE_MANAGER_THREAD_CURRENT_COUNT.getName()),
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR::getPoolSize);
+    MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
+        MetricKey.WORKER_CACHE_MANAGER_THREAD_QUEUE_WAITING_TASK_COUNT.getName()),
```
Comment (jiacheliu3) [inline core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java:81]: For these metrics, pls use cached gauges and at least cache for a few seconds.

File: core/common/src/main/java/alluxio/metrics/MetricKey.java [hunk 11]
```diff
@@ -1340,6 +1340,41 @@ public MetricKey build() {
...
+      new Builder("Worker.CacheMangerThreadMaxCount")
+          .setDescription("The maximum allowed number of block cache "
+              + "thread in the cache manger thread pool")
+          .setMetricType(MetricType.GAUGE)
+          .setIsClusterAggregated(false)
+          .build();
+  public static final MetricKey WORKER_CACHE_MANAGER_COMPLETED_TASK_COUNT =
+      new Builder("Worker.CacheMangerCompleteTaskCount")
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/metrics/MetricKey.java:1372]: typo

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Added a few minor comments, mostly LGTM, thanks!

File: core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java [hunk 12]
```diff
@@ -70,6 +71,22 @@ public final class GrpcExecutors {
...
+    MetricsSystem.registerCachedGaugeIfAbsent(MetricsSystem.getMetricName(
+        MetricKey.WORKER_CACHE_MANAGER_THREAD_ACTIVE_COUNT.getName()),
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR::getActiveCount, 5, TimeUnit.SECONDS);
+    MetricsSystem.registerCachedGaugeIfAbsent(MetricsSystem.getMetricName(
+        MetricKey.WORKER_CACHE_MANAGER_THREAD_CURRENT_COUNT.getName()),
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR::getPoolSize, 5, TimeUnit.SECONDS);
+    MetricsSystem.registerCachedGaugeIfAbsent(MetricsSystem.getMetricName(
+        MetricKey.WORKER_CACHE_MANAGER_THREAD_QUEUE_WAITING_TASK_COUNT.getName()),
+        CACHE_MANAGER_THREAD_POOL_EXECUTOR.getQueue()::size, 5, TimeUnit.SECONDS);
+    MetricsSystem.registerCachedGaugeIfAbsent(MetricsSystem.getMetricName(
```
Comment (jiacheliu3) [inline core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java:83]: nit: add a comment why this is different, like:
```
// This value does not change after the process starts
```

**Review by jiacheliu3 — APPROVED** [review]
LGTM with a minor comment, thanks for the improvement!

File: core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java [hunk 13]
```diff
@@ -38,13 +38,14 @@ public final class GrpcExecutors {
   private static final long THREAD_STOP_MS = Constants.SECOND_MS * 10;
   private static final int THREADS_MIN = 4;
 
-  public static final ExecutorService CACHE_MANAGER_EXECUTOR =
-      new ImpersonateThreadPoolExecutor(new ThreadPoolExecutor(THREADS_MIN,
+  public static final ThreadPoolExecutor CACHE_MANAGER_THREAD_POOL_EXECUTOR =
```
Comment (dbw9580) [inline core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java:41]: This one can be private. See other thread pool executor fields in this class.

**Review by dbw9580 — APPROVED** [review]
LGTM with a minor comment

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Pull request merged**